From cbb1f88638d3f1c981a2633be4ca3f4cf9f443af Mon Sep 17 00:00:00 2001
From: Ali Kianian <ali.kianian@qt.io>
Date: Tue, 7 Oct 2025 12:30:26 +0300
Subject: [PATCH 380/553] QUndoStack: Notify changes when the command is
 obsolete within redo

If a command gets obsolete within the redo method, the index is not
changed, but the status of canRedo, canUndo, redoText, and undoText
might change.
We should report them when any of these changes are met.

Fixes: QTBUG-138567
Pick-to: 6.8
Change-Id: I172a4e0e3df3a3bc40ac6757d79dde7b376ee96a
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 31b0dadb0f371fc94652782c28024f135a0b6f4b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/gui/util/qundostack.cpp                   | 23 ++++-
 src/gui/util/qundostack_p.h                   | 18 ++++
 .../gui/util/qundostack/tst_qundostack.cpp    | 98 ++++++++++++++++++-
 3 files changed, 132 insertions(+), 7 deletions(-)

diff --git a/src/gui/util/qundostack.cpp b/src/gui/util/qundostack.cpp
index fdd5b07a8d4..3d1d8a2b788 100644
--- a/src/gui/util/qundostack.cpp
+++ b/src/gui/util/qundostack.cpp
@@ -419,13 +419,24 @@ void QUndoStackPrivate::setIndex(int idx, bool clean)
 
     bool was_clean = index == clean_index;
 
-    if (idx != index) {
+    const bool indexChanged = idx != index;
+    if (indexChanged) {
         index = idx;
         emit q->indexChanged(index);
-        emit q->canUndoChanged(q->canUndo());
-        emit q->undoTextChanged(q->undoText());
-        emit q->canRedoChanged(q->canRedo());
-        emit q->redoTextChanged(q->redoText());
+    }
+
+    const ActionState newUndoState{q->canUndo(), q->undoText()};
+    if (indexChanged || newUndoState != undoActionState) {
+        undoActionState = newUndoState;
+        emit q->canUndoChanged(undoActionState.enabled);
+        emit q->undoTextChanged(undoActionState.text);
+    }
+
+    const ActionState newRedoState{q->canRedo(), q->redoText()};
+    if (indexChanged || newRedoState != redoActionState) {
+        redoActionState = newRedoState;
+        emit q->canRedoChanged(redoActionState.enabled);
+        emit q->redoTextChanged(redoActionState.text);
     }
 
     if (clean)
@@ -796,6 +807,8 @@ void QUndoStack::redo()
 
         if (d->clean_index > idx)
             resetClean();
+
+        d->setIndex(idx, false);
     } else {
         d->setIndex(d->index + 1, false);
     }
diff --git a/src/gui/util/qundostack_p.h b/src/gui/util/qundostack_p.h
index d418fa48268..fea201ce62d 100644
--- a/src/gui/util/qundostack_p.h
+++ b/src/gui/util/qundostack_p.h
@@ -49,12 +49,30 @@ class QUndoStackPrivate : public QObjectPrivate
 public:
     QUndoStackPrivate() : index(0), clean_index(0), group(nullptr), undo_limit(0) {}
 
+    /*!
+     * \internal
+     * \brief Holds the presentation state for an Undo or Redo command.
+     * This structure serves a change-detection purpose.
+     */
+    struct ActionState
+    {
+        bool enabled = false;
+        QString text;
+
+        bool operator!=(const ActionState &other) const noexcept
+        {
+            return enabled != other.enabled || text != other.text;
+        }
+    };
+
     QList<QUndoCommand*> command_list;
     QList<QUndoCommand*> macro_stack;
     int index;
     int clean_index;
     QUndoGroup *group;
     int undo_limit;
+    ActionState undoActionState;
+    ActionState redoActionState;
 
     void setIndex(int idx, bool clean);
     bool checkUndoLimit();
diff --git a/tests/auto/gui/util/qundostack/tst_qundostack.cpp b/tests/auto/gui/util/qundostack/tst_qundostack.cpp
index 3567bc60975..344003c9328 100644
--- a/tests/auto/gui/util/qundostack/tst_qundostack.cpp
+++ b/tests/auto/gui/util/qundostack/tst_qundostack.cpp
@@ -93,6 +93,18 @@ private:
     QPoint m_newPoint;
 };
 
+class SingleRedoCommand : public QUndoCommand
+{
+public:
+    SingleRedoCommand(QUndoCommand *parent = nullptr);
+    ~SingleRedoCommand();
+
+    void redo() override;
+
+private:
+    bool m_firstRun = true;
+};
+
 InsertCommand::InsertCommand(QString *str, int idx, const QString &text,
                             QUndoCommand *parent)
     : QUndoCommand(parent)
@@ -255,6 +267,22 @@ bool MoveMouseCommand::mergeWith(const QUndoCommand *other)
     return true;
 }
 
+SingleRedoCommand::SingleRedoCommand(QUndoCommand *parent)
+    : QUndoCommand(parent)
+{
+    setText("single redo");
+}
+
+SingleRedoCommand::~SingleRedoCommand() = default;
+
+void SingleRedoCommand::redo()
+{
+    if (m_firstRun)
+        m_firstRun = false;
+    else
+        setObsolete(true);
+}
+
 /******************************************************************************
 ** tst_QUndoStack
 */
@@ -2637,11 +2665,77 @@ void tst_QUndoStack::obsolete()
                 "",           // undoText
                 false,        // canRedo
                 "",           // redoText
-                true,        // cleanChanged
+                true,         // cleanChanged
                 false,        // indexChanged
                 false,        // undoChanged
                 false);       // redoChanged
 
+    stack.push(new SingleRedoCommand());
+    checkState(redoTextChangedSpy,
+               canRedoChangedSpy,
+               undoTextChangedSpy,
+               redoAction,
+               undoAction,
+               canUndoChangedSpy,
+               cleanChangedSpy,
+               indexChangedSpy,
+               stack,
+               false,         // clean
+               1,             // count
+               1,             // index
+               true,          // canUndo
+               "single redo", // undoText
+               false,         // canRedo
+               "",            // redoText
+               false,         // cleanChanged
+               true,          // indexChanged
+               true,          // undoChanged
+               true);         // redoChanged
+
+    stack.undo();
+    checkState(redoTextChangedSpy,
+               canRedoChangedSpy,
+               undoTextChangedSpy,
+               redoAction,
+               undoAction,
+               canUndoChangedSpy,
+               cleanChangedSpy,
+               indexChangedSpy,
+               stack,
+               false,         // clean
+               1,             // count
+               0,             // index
+               false,         // canUndo
+               "",            // undoText
+               true,          // canRedo
+               "single redo", // redoText
+               false,         // cleanChanged
+               true,          // indexChanged
+               true,          // undoChanged
+               true);         // redoChanged
+
+    stack.redo();
+    checkState(redoTextChangedSpy,
+               canRedoChangedSpy,
+               undoTextChangedSpy,
+               redoAction,
+               undoAction,
+               canUndoChangedSpy,
+               cleanChangedSpy,
+               indexChangedSpy,
+               stack,
+               false,         // clean
+               0,             // count
+               0,             // index
+               false,         // canUndo
+               "",            // undoText
+               false,         // canRedo
+               "",            // redoText
+               false,         // cleanChanged
+               false,         // indexChanged
+               false,         // undoChanged
+               true);         // redoChanged
+
     stack.push(new MoveMouseCommand(&mouse, mouse, QPoint(0, 0))); // #1 should not merge but will be deleted (b/c oldPoint == newPoint)
     QCOMPARE(mouse, QPoint(0, 0));
     checkState(redoTextChangedSpy,
@@ -3008,7 +3102,7 @@ void tst_QUndoStack::obsolete()
                 false,        // cleanChanged
                 false,        // indexChanged
                 false,        // undoChanged
-                false);       // redoChanged
+                true);        // redoChanged
     QCOMPARE(stack.cleanIndex(), -1);
 
     stack.redo();
-- 
2.51.2

